[DebugInfo] Change return type of DWARFDebugAbbrev::parse

bulbazord opened the pull request:
To make DWARFDebugAbbrev more amenable to error-handling, I would like to change the return type of DWARFDebugAbbrev::parse from `void` to `Error`. Users of DWARFDebugAbbrev can consume the error if they want to use all the valid DWARF that was parsed (without worrying about the malformed DWARF) or stop when the parse fails if the use case needs to be strict.

This also will bring the LLVM DWARFDebugAbbrev interface closer to LLDB's which opens up the opportunity for LLDB adopt the LLVM implementation with minimal changes.

bulbazord committed:
[DebugInfo] Change return type of DWARFDebugAbbrev::parse

To make DWARFDebugAbbrev more amenable to error-handling, I would like
to change the return type of DWARFDebugAbbrev::parse from `void` to
`Error`. Users of DWARFDebugAbbrev can consume the error if they want to
use all the valid DWARF that was parsed or stop when the parse fails if
the use case needs to be strict.

This also will bring the LLVM DWARFDebugAbbrev interface closer to
LLDB's which opens up the opportunity for LLDB adopt the LLVM
implementation with minimal changes.

jh7370 reviewed (commented):
In principle, seems reasonable to me, though I worry slightly about the impact on existing users who won't expect this breaking-if-they-have-appropriate-options-selected change. How confident are you that you've caught all in-tree uses?

Also, does this change allow us to check the lower-level error in a test somewhere, e.g. in a unit test?

bulbazord committed:
[DebugInfo] Add unit test for DWARFDebugAbbrev::parse

Just to make sure that an error is correctly propagated upwards.

bulbazord commented:
> In principle, seems reasonable to me, though I worry slightly about the impact on existing users who won't expect this breaking-if-they-have-appropriate-options-selected change. How confident are you that you've caught all in-tree uses?

I would say I lean more towards confident than not. I've built clang, lld, lldb, bolt, and compiler-rt (all on an arm64-apple-macOS machine). The call-sites that I updated were found through a combination of commenting out the method in the header to introduce a compilation failure, IDE tooling, and tests. If there are places I did not catch, it is because of conditional compilation checks that my build did not meet. As always, if and when I land this I will pay attention to the CI machines.

I assume downstream users will adjust accordingly.

> Also, does this change allow us to check the lower-level error in a test somewhere, e.g. in a unit test?

Added a unit test similar to the other ones I introduced.

bulbazord committed:
Adjust formatting of new unittest

jh7370 reviewed (approved):
LGTM.
